### PR TITLE
[4.0] replace usage of curl in python with urllib

### PR DESCRIPTION
**Pull request by ClaytonCalabrese**
Replaces all instances of curl in the tests with urllib. This is done via a function that replaces `processCurlCmd` and calls into urllib.

This should result in increased efficiency because we aren't relying on an outside command line to execute things.

### CODE REVIEW

**Review by jgiszczak — APPROVED** [review]

**Comment by jgiszczak** [comment]
The merge currently targets `main` but doesn't seem to be based on it.

**Comment by heifner** [comment]
The `ubuntu20-nodeos_startup_catchup_lr_test` is failing with:
```
error 2022-11-07T17:02:28.146 nodeos    beast_http_session.hpp:313    handle_exception     ] fc::exception: 7 bad_cast_exception: Bad Cast
Invalid cast from string_type to Array
    {"type":"string_type"}
    nodeos  variant.cpp:549 get_array
```

File: tests/TestHarness/Node.py [hunk 1]
```diff
@@ -1017,20 +1020,29 @@ def killNodeOnProducer(self, producer, whereInSequence, blockType=BlockType.head
...
-    def processCurlCmd(self, resource, command, payload, silentErrors=True, exitOnError=False, exitMsg=None, returnType=ReturnType.json):
-        cmd="curl %s/v1/%s/%s -d '%s' -X POST -H \"Content-Type: application/json\"" % \
-            (self.endpointHttp, resource, command, payload)
+        payload={ "producer":producer, "where_in_sequence":whereInSequence, "based_on_lib":basedOnLib }
+        return self.processUrllibRequest("test_control", "kill_node_on_producer", payload, silentErrors=silentErrors, exitOnError=exitOnError, exitMsg=exitMsg, returnType=returnType)
+
+    def processUrllibRequest(self, resource, command, payload={}, silentErrors=False, exitOnError=False, exitMsg=None, returnType=ReturnType.json, endpoint=None):
+        if not endpoint:
+            endpoint = self.endpointHttp
+        cmd = "%s/v1/%s/%s" % (endpoint, resource, command)
```
Comment (oschwaldp-oci) [inline tests/TestHarness/Node.py:1029]: Should this be an f string?

File: tests/plugin_http_api_test.py [hunk 2]
```diff
@@ -8,20 +8,18 @@
...
 import re
 
-from TestHarness import Account, Node, TestHelper, Utils, WalletMgr
+from TestHarness import Account, Node, TestHelper, Utils, WalletMgr, ReturnType
 
 class PluginHttpTest(unittest.TestCase):
     sleep_s = 2
-    base_node_cmd_str = ("curl http://%s:%s/v1/") % (TestHelper.LOCAL_HOST, TestHelper.DEFAULT_PORT)
-    base_wallet_cmd_str = ("curl http://%s:%s/v1/") % (TestHelper.LOCAL_HOST, TestHelper.DEFAULT_WALLET_PORT)
+    base_wallet_cmd_str = ("http://%s:%s") % (TestHelper.LOCAL_HOST, TestHelper.DEFAULT_WALLET_PORT)
```
Comment (oschwaldp-oci) [inline tests/plugin_http_api_test.py:15]: Could probably use an f string here as well.

File: tests/plugin_http_api_test.py [hunk 3]
```diff
@@ -129,1385 +124,1227 @@ def test_ChainApi(self) :
...
         # abi_json_to_bin with valid parameter
-        valid_cmd = ("%s%s '{%s,%s,%s}'") % (default_cmd,
-                                             self.http_post_str,
-                                             "\"code\":\"eosio.token\"",
-                                             "\"action\":\"issue\"",
-                                             "\"args\":{\"to\":\"eosio.token\", \"quantity\":\"1.0000\%20EOS\",\"memo\":\"m\"}")
-        ret_json = Utils.runCmdReturnJson(valid_cmd)
+        payload = {"code":"eosio.token",
+                     "action":"issue",
+                     "args":{"to":"eosio.token", "quantity":"1.0000\%20EOS","memo":"m"}}
```
Comment (oschwaldp-oci) [inline tests/plugin_http_api_test.py:549]: Could we line up the indentation? Line up opening quotes.

File: tests/plugin_http_api_test.py [hunk 4]
```diff
@@ -129,1385 +124,1227 @@ def test_ChainApi(self) :
...
         # abi_bin_to_json with valid parameter
-        valid_cmd = ("%s%s '{%s,%s,%s}'") % (default_cmd,
-                                             self.http_post_str,
-                                             "\"code\":\"eosio.token\"",
-                                             "\"action\":\"issue\"",
-                                             "\"args\":\"ee6fff5a5c02c55b6304000000000100a6823403ea3055000000572d3ccdcd0100000000007015d600000000a8ed32322a00000000007015d6000000005c95b1ca102700000000000004454f53000000000968656c6c6f206d616e00\"")
-        ret_json = Utils.runCmdReturnJson(valid_cmd)
+        payload = {"code":"eosio.token",
+                     "action":"issue",
+                     "args":"ee6fff5a5c02c55b6304000000000100a6823403ea3055000000572d3ccdcd0100000000007015d600000000a8ed32322a00000000007015d6000000005c95b1ca102700000000000004454f53000000000968656c6c6f206d616e00"}
```
Comment (oschwaldp-oci) [inline tests/plugin_http_api_test.py:569]: Line up opening quotes.

File: tests/plugin_http_api_test.py [hunk 5]
```diff
@@ -129,1385 +124,1227 @@ def test_ChainApi(self) :
...
-                                                   self.http_post_str,
-                                                   "\"signatures\":[\"SIG_K1_KeqfqiZu1GwUxQb7jzK9Fdks6HFaVBQ9AJtCZZj56eG9qGgvVMVtx8EerBdnzrhFoX437sgwtojf2gfz6S516Ty7c22oEp\"]",
-                                                   "\"compression\": true",
-                                                   "\"packed_context_free_data\": \"context_free_data\"",
-                                                   "\"packed_trx\": \"packed_trx\"")
-        ret_json = Utils.runCmdReturnJson(valid_cmd)
+        payload = {"signatures":["SIG_K1_KeqfqiZu1GwUxQb7jzK9Fdks6HFaVBQ9AJtCZZj56eG9qGgvVMVtx8EerBdnzrhFoX437sgwtojf2gfz6S516Ty7c22oEp"],
+                     "compression": "true",
+                     "packed_context_free_data": "context_free_data",
+                     "packed_trx": "packed_trx"}
```
Comment (oschwaldp-oci) [inline tests/plugin_http_api_test.py:663]: Line up opening quotes.

File: tests/plugin_http_api_test.py [hunk 6]
```diff
@@ -129,1385 +124,1227 @@ def test_ChainApi(self) :
...
+                     "ref_block_num": 34881,
+                     "ref_block_prefix":2972818865,
+                     "max_net_usage_words":0,
+                     "max_cpu_usage_ms":0,
+                     "delay_sec":0,
+                     "context_free_actions":[],
+                     "actions":[{"account":"eosio.token","name": "transfer","authorization": [{"actor": "han","permission": "active"}],"data": "000000000000a6690000000000ea305501000000000000000453595300000000016d"}],
+                     "transaction_extensions": [],
+                     "signatures": ["SIG_K1_KeqfqiZu1GwUxQb7jzK9Fdks6HFaVBQ9AJtCZZj56eG9qGgvVMVtx8EerBdnzrhFoX437sgwtojf2gfz6S516Ty7c22oEp"],
+                     "context_free_data": []}
```
Comment (oschwaldp-oci) [inline tests/plugin_http_api_test.py:624]: Line up opening quotes.

File: tests/plugin_http_api_test.py [hunk 7]
```diff
@@ -129,1385 +124,1227 @@ def test_ChainApi(self) :
...
+        payload = {"ref_block_num":"100",
+                     "ref_block_prefix": "137469861",
+                     "expiration": "2020-09-25T06:28:49",
+                     "scope":["initb", "initc"],
+                     "actions": [{"code": "currency","type":"transfer","recipients": ["initb", "initc"],"authorization": [{"account": "initb", "permission": "active"}],"data":"000000000041934b000000008041934be803000000000000"}],
+                     "signatures": [],
+                     "authorizations": [],
+                     "available_keys":["EOS4toFS3YXEQCkuuw1aqDLrtHim86Gz9u3hBdcBw5KNPZcursVHq",
+                     "EOS7d9A3uLe6As66jzN8j44TXJUqJSK3bFjjEEqR4oTvNAB3iM9SA",
+                     "EOS6MRyAjQq8ud7hVNYcfnVPJqcVpscN5So8BhtHuGYqET5GDW5CV"]}
```
Comment (oschwaldp-oci) [inline tests/plugin_http_api_test.py:596]: Line up opening quotes.

File: tests/plugin_http_api_test.py [hunk 8]
```diff
@@ -129,1385 +124,1227 @@ def test_ChainApi(self) :
...
+                              "max_cpu_usage_ms":0,
+                              "delay_sec":0,
+                              "context_free_actions":[],
+                              "actions": [],
+                              "transaction_extensions": [],
+                              "signatures": ["SIG_K1_KeqfqiZu1GwUxQb7jzK9Fdks6HFaVBQ9AJtCZZj56eG9qGgvVMVtx8EerBdnzrhFoX437sgwtojf2gfz6S516Ty7c22oEp"],
+                              "context_free_data": []}
+        payload = [signed_transaction,
+                     ["EOS696giL6VxeJhtEgKtWPK8aQeT8YXNjw2a7vE5wHunffhfa5QSQ"],
+                     "cf057bbfb72640471fd910bcb67639c22df9f92470936cddc1ade0e2f2e7dc4f"]
```
Comment (oschwaldp-oci) [inline tests/plugin_http_api_test.py:1102]: Maybe line things up here as well?

**Review by oschwaldp-oci — COMMENTED** [review]
LGTM.  A couple stylistic comments/recommendations.

File: tests/TestHarness/Node.py (hunk 1 above)
Comment (jgiszczak) [reply tests/TestHarness/Node.py:1029]: I think so.  They really are so much more readable.  It did make it easy to inspect the diff, but that's only momentarily useful.

**Review by jgiszczak — APPROVED** [review]

**Review by oschwaldp-oci — APPROVED** [review]

**Pull request merged**